### PR TITLE
[SLP]Consider (f)sub, being operand of llvm.(f)abs/icmp eq/ne 0, commutative.

### PR DESCRIPTION
**Pull request by alexey-bataev**
If (f)sub is only operand of llvm.(f)abs or icmp eq/ne 0 (int only), we can consider it as commutative operation, just need to drop wrapping flags for ineteger
operation.

https://alive2.llvm.org/ce/z/GxvxjB for correctness of abs with dropped
flags.

### CODE REVIEW

File: llvm/lib/Transforms/Vectorize/SLPVectorizer.cpp [hunk 1]
```diff
@@ -305,7 +305,32 @@ static bool isCommutative(Instruction *I) {
...
+                     return IC && IC->getIntrinsicID() == Intrinsic::abs;
+                   })) ||
+           (BO->getOpcode() == Instruction::FSub &&
+            !BO->hasNUsesOrMore(UsesLimit) &&
+            all_of(BO->uses(), [](const Use &U) {
+              if (U.getOperandNo() != 0)
+                return false;
+              const auto *IC = dyn_cast<IntrinsicInst>(U.getUser());
+              return IC && IC->getIntrinsicID() == Intrinsic::fabs;
+            }));
```
Comment (RKSimon) [inline llvm/lib/Transforms/Vectorize/SLPVectorizer.cpp:333]: Yet more code that really shouldn't be inside SLP - can we find somewhere more general to put it please?

@dtcxzyw do you know where we can move this, and do you know any similar folds in the middle-end that we can share?

Comment (alexey-bataev) [reply llvm/lib/Transforms/Vectorize/SLPVectorizer.cpp:333]: > Yet more code that really shouldn't be inside SLP - can we find somewhere more general to put it please?
 Actually, it should be in SLP. It limits the graph and thus affects the cost. It does not handle all possible cases, though (e.g., operands with same opcodes can be non-schedulable, we cannot detect it here, maybe later), so having something in the instcombine is still good.

File: llvm/test/Transforms/PhaseOrdering/X86/vector-reductions.ll [hunk 2]
```diff
@@ -63,7 +63,7 @@ define i32 @TestVectorsEqual(ptr noalias %Vec0, ptr noalias %Vec1, i32 %Toleranc
 ; CHECK-NEXT:  entry:
 ; CHECK-NEXT:    [[TMP0:%.*]] = load <4 x i32>, ptr [[VEC0:%.*]], align 4
 ; CHECK-NEXT:    [[TMP1:%.*]] = load <4 x i32>, ptr [[VEC1:%.*]], align 4
-; CHECK-NEXT:    [[TMP2:%.*]] = sub nsw <4 x i32> [[TMP0]], [[TMP1]]
+; CHECK-NEXT:    [[TMP2:%.*]] = sub <4 x i32> [[TMP0]], [[TMP1]]
```
Comment (RKSimon) [inline llvm/test/Transforms/PhaseOrdering/X86/vector-reductions.ll:66]: Losing the NSW prevents the ABDS pattern

File: llvm/lib/Transforms/Vectorize/SLPVectorizer.cpp (hunk 1 above)
Comment (RKSimon) [reply llvm/lib/Transforms/Vectorize/SLPVectorizer.cpp:333]: OK.

File: llvm/test/Transforms/PhaseOrdering/X86/vector-reductions.ll (hunk 2 above)
Comment (alexey-bataev) [reply llvm/test/Transforms/PhaseOrdering/X86/vector-reductions.ll:66]: Could you point me what is this pattern?

Comment (RKSimon) [reply llvm/test/Transforms/PhaseOrdering/X86/vector-reductions.ll:66]: DAGCombiner::foldABSToABD

Comment (alexey-bataev) [reply llvm/test/Transforms/PhaseOrdering/X86/vector-reductions.ll:66]: Thanks, will relax the check and do best work with flags.

**Comment by alexey-bataev** [comment]
Ping!

**Comment by alexey-bataev** [comment]
Ping!

**Comment by alexey-bataev** [comment]
Ping!

**Comment by alexey-bataev** [comment]
Ping!

**Comment by alexey-bataev** [comment]
Ping!

**Review by RKSimon — APPROVED** [review]
LGTM

**Pull request merged**